src/PublicAllocator.sol

```diff
@@ -1,25 +1,127 @@
+        if (vault == address(0)) revert ErrorsLib.ZeroAddress();
+        OWNER = newOwner;
+        VAULT = IMetaMorpho(vault);
+        MORPHO = VAULT.MORPHO();
     }
 
     /// PUBLIC ///
 
-    function reallocate(MarketAllocation[] calldata allocations) external {
-        vault.reallocate(allocations);
+    function reallocate(MarketAllocation[] calldata allocations) external payable {
+        if (msg.value != fee) revert ErrorsLib.IncorrectFee(msg.value);
+
+        uint256[] memory assets = new uint256[](allocations.length);
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            // Do not compute interest twice for every market
+            MORPHO.accrueInterest(allocations[i].marketParams);
+            assets[i] = MORPHO.expectedSupplyAssets(allocations[i].marketParams, address(VAULT));
+        }
+
+        VAULT.reallocate(allocations);
+
+        MarketParams memory marketParams;
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            marketParams = allocations[i].marketParams;
+            Id id = marketParams.id();
+            uint256 newAssets = MORPHO.expectedSupplyAssets(marketParams, address(VAULT));
+            if (newAssets >= assets[i]) {
+                if (newAssets > supplyCap[id]) revert ErrorsLib.PublicAllocatorSupplyCapExceeded(id);
+                uint128 inflow = (newAssets - assets[i]).toUint128();
+                flowCap[id].maxIn -= inflow;
+                flowCap[id].maxOut = (flowCap[id].maxOut).saturatingAdd(inflow);
+            } else {
+                uint128 outflow = (assets[i] - newAssets).toUint128();
+                flowCap[id].maxIn = (flowCap[id].maxIn).saturatingAdd(outflow);
+                flowCap[id].maxOut -= outflow;
+            }
+        }
+
+        emit EventsLib.PublicReallocate(msg.sender, msg.value);
+    }
+
+    /// OWNER ONLY ///
+
+    function setFee(uint256 _fee) external onlyOwner {
+        if (fee == _fee) revert ErrorsLib.AlreadySet();
+        fee = _fee;
+        emit EventsLib.SetFee(_fee);
+    }
+
+    function transferFee(address payable feeRecipient) external onlyOwner {
+        uint256 balance = address(this).balance;
+        feeRecipient.transfer(balance);
+        emit EventsLib.TransferFee(balance);
+    }
+
+    // Set flow cap
+    // Doesn't revert if it doesn't change the storage at all
+    function setFlowCaps(FlowConfig[] calldata flowCaps) external onlyOwner {
+        for (uint256 i = 0; i < flowCaps.length; ++i) {
+            flowCap[flowCaps[i].id] = flowCaps[i].cap;
+        }
+
+        emit EventsLib.SetFlowCaps(flowCaps);
+    }
+
+    // Set supply cap. Public reallocation will not be able to increase supply if it ends above its cap.
+    // Doesn't revert if it doesn't change the storage at all
+    function setSupplyCaps(SupplyConfig[] calldata supplyCaps) external onlyOwner {
+        for (uint256 i = 0; i < supplyCaps.length; ++i) {
```

Comment:
I think that this is no longer an opti
```suggestion
        for (uint256 i = 0; i < supplyCaps.length; i++) {
```

Reply:
d7b76268069f5c04da7bef8c93c76fbca278d88b